### PR TITLE
Fix for text echo issue when running sbt.

### PR DESCRIPTION
**Pull request by kczulko**
Fixes https://github.com/sbt/sbt/issues/3453

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @kczulko,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3507/head&deploy_to_bintray=false">this link.</A>

**Comment by kczulko** [comment]
Hi People :)

This is my very first sbt contribution. Since this lack of echo was pissing me off and I found the unsolved issue for that, I'd played a little bit around sbt source code and made this dummy fix. I'm almost sure it is not the right way to resolve that ( :) ) but I thought it would be great to start the discussion here.

Best regards,
Karol

**Comment by lightbend-cla-validator** [comment]
Hi @kczulko,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3507/head&deploy_to_bintray=false">this link.</A>

**Comment by eed3si9n** [comment]
Hi Karol, welcome to sbt project. I really appreciate the "let me fix it myself" spirit.

File: internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala [hunk 1]
```diff
@@ -120,7 +120,10 @@ private[sbt] object JLine {
   def usingTerminal[T](f: jline.Terminal => T): T =
     withTerminal { t =>
       t.restore
-      f(t)
+      val result = f(t)
+      // UnixTerminal::init call removes echo
+      t.setEchoEnabled(true)
```
Comment (eed3si9n) [inline internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala]: Could you try `t.restore` here instead?
I am guessing that would bring the state back to whatever the terminal was before.

**Review by eed3si9n — CHANGES_REQUESTED** [review]

File: internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala (hunk 1 above)
Comment (kczulko) [reply internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala]: Thx @eed3si9n. Hope to check that today afternoon. Will give feedback here. Best regards :)

**Comment by lightbend-cla-validator** [comment]
Hi @kczulko,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3507/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by kczulko** [comment]
@eed3si9n so before returning a result I called `t.restore` and as you wrote, that also fixed the problem. Can you please guide me whether I can put some test for that? Would be great to eliminate that behavior once and for all. Thx in advance :)

**Comment by eed3si9n** [comment]
> Can you please guide me whether I can put some test for that? Would be great to eliminate that behavior once and for all. Thx in advance :)

I am not sure. Is there a way to programmatically observe that the terminal is echoing, in somewhat cross platform way?

File: internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala [hunk 2]
```diff
@@ -119,8 +119,9 @@ private[sbt] object JLine {
    */
   def usingTerminal[T](f: jline.Terminal => T): T =
     withTerminal { t =>
+      val result = f(t)
```
Comment (eed3si9n) [inline internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala]: The Javadoc comment says "This ensures synchronized access as well as re-enabling echo after getting the Terminal.", so I am guessing that you'd need to call `t.restore` both before and after `f(t)`.

**Review by eed3si9n — CHANGES_REQUESTED** [review]

**Comment by lightbend-cla-validator** [comment]
Hi @kczulko,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by kczulko** [comment]
@eed3si9n,
Ok. I thought that the root cause of that issue was calling `t.restore` before returning a value. I've found that `usingTerminal` function has four usages. Three of them are taking height/widht of the terminal and one of them (`LineReader::createReader`) ignores terminal (`t`) parameter value. I've fixed that in the way you want so now `t.restore` is called both before and after `f(t)` call.

**Review by eed3si9n — APPROVED** [review]

**Comment by eed3si9n** [comment]
LGTM. Thanks for the contribution.

**Comment by egp** [comment]
Hooray. Thank you so much!

**Pull request merged**